pkg/kube/client.go

```diff
@@ -44,16 +54,50 @@ const (
 	defaultLocalAddress = "localhost"
 )
 
-// Client is a helper for common Kubernetes client operations
-type Client interface {
+// BaseClient contains the core set of interfaces shared between all client users. This allows us to split
+// functions used only by CLI tools (like istioctl) and functions used only by servers (like Informers)
+type BaseClient interface {
+	// TODO - stop embedding this, it will conflict with future additions. Use Kube() instead is preferred
+	// TODO - add istio/client-go and service-apis
 	kubernetes.Interface
-	rest.Interface
+
+	// Kube returns the core kube client
+	Kube() kubernetes.Interface
+
+	// DynamicClient returns the Dynamic kube client.
+	DynamicClient() dynamic.Interface
+
+	// MetadataClient returns the Metadata kube client.
+	MetadataClient() metadata.Interface
 
 	// Config returns the Kubernetes rest config used to configure the clients.
 	Config() *rest.Config
 
 	// Ext returns the API extensions client.
 	Ext() kubeExtClient.Interface
+}
+
+// Client is a helper for common Kubernetes client operations
+type Client interface {
+	BaseClient
+
+	// KubeInformer returns an informer for core kube client
+	KubeInformer() informers.SharedInformerFactory
+
+	// DynamicInformer returns an informer for dynamic client
+	DynamicInformer() dynamicinformer.DynamicSharedInformerFactory
+
+	// MetadataInformer returns an informer for metadata client
+	MetadataInformer() metadatainformer.SharedInformerFactory
+
+	// RunAndWait starts all informers and waits for their caches to sync.
+	// Warning: this must be called AFTER .Informer() is called, which will register the informer.
+	RunAndWait(stop <-chan struct{})
+}
+
+// IstioctlClient is an extended client with additional helpers/functionality specific to Istioctl
+type IstioctlClient interface {
```

Comment:
I thought we were going to keep it to 2 interfaces for now? There are things here that aren't istioctl specific (port-forwarding, exec, etc.). It might be easier to split out the istioctl-specific API (and actually keep it under istioctl) after this lands.

Reply:
I tried this and struggled to do it cleanly; the two require info that is unique to both (istioctl doesn't want informers, istioctl stuff requires things like revision so I did it in this one as the changes to do so was actually fairly small. Is it just a naming issue or do you have concerns with the structure itself? It could be ExtendedClient or something, I just don't want server code doing dumb things like port forwarding or exec

Reply:
ExtendedClient works.